docs/RFCS/20201020_sql_query_denylist.md

```diff
@@ -0,0 +1,183 @@
+
+To confirm, they can add a staging denylist to the location they had specified for the staging denylist and execute the query. The query will execute, and if the query matches the regex query pattern specified in the denylist, the administrator will be able to view this in the Admin UI, SQL console, and CRDB logs. 
+
+The administrator will then be able to add a denylist to the denylist location specified. Any queries after this moment will be matched against the denylist, and the denial logged for the administrator if there is a match. A time-series graph of queries denied, and queries that would be denied in the case of a staging denylist, will also be available to view on the Admin UI.
+
+The administrator may modify the file content as needed; and if the entire denylist is no longer available, they may remove it entirely. The changes will take effect immediately.
+
+### User experience from a developer's point of view
+A developer may be sending a query that has been denied on the denylist or would be denied and is currently on the staging denylist. In the latter case, they will also be able to view the warning on console that this query is still executing, but would be denied due to a pattern match on a staging denylist. In the former case, they will receive an error message that "query matched a pattern in the denylist by the database administrator" with the specific regex query pattern that it matched. 
+
+# Reference-level explanation
+
+## Detailed design
+
+### Design overview
+We will implement a file-managed denylist, stored locally on each machine as a YAML configuration file. The exact location of the denylist should be configured with a flag – this will allow users to specify the location of their file, such as via the k8s ConfigMap for CockroachCloud SREs. The file location will be watched so that as soon as the config file is created, modified, or removed, the denylist will be updated. Pattern matching between queries executed and queries on the denylist will be done via regex so users maintain granular control of the queries affected by the denylist.
+
+A staging YAML file will be supported so users will be able to confirm the queries that would be denied: when the staging file is set, users will see logs of the queries that are not yet denied but will be denied if the non-staging denylist file is set.  
+
+### Configuring denylist location
+A denylist location flag can be configured in [pkg/cli/flags.go](https://github.com/cockroachdb/cockroach/blob/master/pkg/cli/flags.go). eg. `--denylist=’denylist.yaml’`. If this flag has not been set, then there will be no location to watch for a denylist and therefore the functionality is off by default. 
```

Comment:
Why not watch a default location?

Reply:
Good point. I had thought there might be some overhead added in having `fsnotify` on and watching for a user who may never use this feature. @otan do you happen to know whether there is? If it's minimal, you're right in that having a default location is a better user experience.